README.md
# WebKit Input Range Fill Lower Hack

Customize the native input range controls like the ```::-ms-fill-lower``` property on WebKit browsers

[![License](http://img.shields.io/:license-mit-blue.svg?style=flat-square)](http://badges.mit-license.org)

![Screenshot](samples.png)

## New Version 3.x Notice
Because of the increasing interest in this "library", I decided to completely rewrite it.

**WARNING**: the old version isn't compatible with this one. If you want to use the older version you can download it from [here]()


## Key Features
* **Uses native HTML input range type (more accessibility, native behavior)**
* **Doesn't require anymore the external CSS** *new version*
* **Multiple elements** *new version*
* **Customizable styles (solid color or gradient)** *new version*
* **Written in TypeScript** *new version*
* **Pure / vanilla JavaScript (no jQuery)**
* **Lightweight** *(3.6kB minified ~1.3kB gzipped)*

View the sample page: https://samuelcarreira.github.io/webkit-input-range-fill-lower

## Quick Usage
You only need to follow 4 simple steps to start using this module:

1. Download the script file from the [dist](dist) folder or [releases](https://github.com/samuelcarreira/webkit-input-range-fill-lower/releases) (minified file: `webkit-input-range-fill-lower.min.js`)
2. Add the input range controls on your web page. **Each control needs a unique ID and the max value**
  ```
  // valid
  <input type="range" min="0" max="10" value="10" id="your-custom-id">
  
  // invalid: no ID
  <input type="range" min="0" max="10" value="10">

  // invalid: no max value
  <input type="range" id="your-custom-id">
  ```
3. Call the library after the page is loaded
  ```
  document.addEventListener('DOMContentLoaded', () => {
    const myRanges = new WebkitInputRangeFillLower({selectors: ['your-custom-id', 'your-custom-id2'], angle: 90, gradient: 'rgba(238,174,202,1) 0%, rgba(198,180,216,1) 74%, rgba(148,187,233,1) 100%'});
  });
  ```
4. (Optional) You can customize the input element on the page stylesheet. View the `index.html` or the [sample page](https://samuelcarreira.github.io/webkit-input-range-fill-lower) source code to learn more

### WebkitInputRangeFillLower(options)

### Class options
Property         | Type     | Required    | Description
---------------- | -------- | ---------- | ----------------------
`selectors`  | `string[]` | yes     | List of HTML ID attributes. If you want to select multiple selectors use an array like `['your-custom-id', 'your-custom-id2']`, single selector can be defined with a simple string `selectors: 'your-custom-id'`
`angle`  	 | `number` | no   | Gradient angle (0-360 degrees). Default value: 90 (only in gradient color)
`color`  	 | `string` | no | CSS solid color. With can use any CSS color format: hexadecimal, description, rgb or rgba
`gradient`  	 | `string` | no   | CSS gradient color definition. The default value is the gradient `#0ABFBC, #FC354C`. You can set the position of the color on the gradient like `rgba(238,174,202,1) 0%, rgba(198,180,216,1) 74%, rgba(148,187,233,1) 100%`

## Compatibility
As the name indicates, this "library" was written to be compatible with the WebKit engine browsers (Google Chrome, Opera, New Microsoft Edge, etc.).
On Mozilla Firefox you cannot expect the same results even with the `-moz-range-progress` added (read more [here](https://developer.mozilla.org/en-US/docs/Web/CSS/::-moz-range-progress) )

![compatibility](compatibility.png)


## Motivation and history
I've created this "hack" because I wanted to use a range slider on my Electron App (a music player with volume control). After I tried some plugins like the [rangeslider.js](http://rangeslider.js.org/) I decided to write something more lightweight and more easy to use.
As Electron apps use Chrome (WebKit) as the render engine, ~~I didn't add any compatibility with other browsers/fallbacks to older versions~~ (on the new 3.0 version I've added the Mozilla firefox and Microsoft Internet Explorer 'fallback' property).

## How this 'hack' works
This 'hack' uses the `::-webkit-slider-runnable-track` CSS pseudo-element (which represents the "track" or the groove in which the indicator slides of an input type="range"). Because you cannot access this property directly on an inline CSS style, this library needs to write a custom style tag directly on the page header where it's set the custom styles and the background style.
Because you can "set the size" of the background (when you use a linear gradient), this 'library' calculates the percentage of the value on the input range element and adjusts the size of the background to that percentage. Each time the input range value is changed, the event listener updates the background-size.


## Credits
- Developed by: Samuel Carreira
- Inspired on this article: http://brennaobrien.com/blog/2014/05/style-input-type-range-in-every-browser.html

## License
- **[MIT license](http://opensource.org/licenses/mit-license.php)**
- Copyright 2020 © <a href="http://samuelcarreira.com" target="_blank">Samuel Carreira</a>.


CHANGELOG.md
2017-04-23 - First public version

2019-04-07 - Version 2.0

2020-04-20 - Version 3.0

src/old.js
/*!
 * Webkit-input-range-fill-lower
 * Version 1.0 (old version)
 *
 * Licensed under MIT
 * Copyright (c) 2020 [Samuel Carreira]
 */

"use strict";

const DEBUG = true;

var inlineStyle = document.createElement('style');
var rangeSelector = document.querySelectorAll('[type=range]');
var inlineStyleContent = new Array;

document.body.appendChild(inlineStyle);

var eventname = new Event('input');

for (let item of rangeSelector) {
    item.addEventListener('input', function () {
        let rangeInterval = Number(this.getAttribute('max') - this.getAttribute('min'));
        let rangePercent = (Number(this.value) + Math.abs(this.getAttribute('min'))) / rangeInterval * 100;

        DEBUG ? console.log("#" + this.id + ": " + rangePercent + "%") : ''; // for debug

        writeStyle({
            id: this.id,
            percent: rangePercent
        });
    }, false);

    item.dispatchEvent(eventname); // update bars at startup
}

/**
 * Write Style element
 * 
 * @param {object} obj id: HTML id, percent: value
 */
function writeStyle(obj) {
    var find = inlineStyleContent.map(x => x.id).indexOf(obj.id);
    var styleText = "";

    if (find === -1) {
        inlineStyleContent.push(obj)
    } else {
        inlineStyleContent[find] = obj;
    }

    for (let item of inlineStyleContent) {
        styleText += '#' + item.id + '::-webkit-slider-runnable-track{background-size:' + item.percent + '% 100%} ';
    }

    inlineStyle.textContent = styleText;
}

src/index.ts
/*!
 * Webkit-input-range-fill-lower
 *
 * Licensed under MIT
 * Copyright (c) 2020 [Samuel Carreira]
 */

interface WebkitInputRangeFillLowerCachedValues {
  /**
   * HTML ID attribute
   */
  id: string;
  /**
   * Range value (0-100%)
   */
  percent: number;
}

interface WebkitInputRangeFillLowerOptions {
  /**
   * List of HTML ID attributes to work with
   */
  selectors: string[];
  /**
   * Gradient angle (0-360 degrees)
   */
  angle?: number;
  /**
   * CSS color
   * @example
   * '#fff'
   * 'rgb(255, 255, 2, 0.5)'
   * 'rgb(128, 125, 23)'
   * 'red'
   */
  color?: string;
  /**
   * Gradient CSS color
   * @example
   * 'rgba(238,174,202,1) 0%, rgba(198,180,216,1) 74%, rgba(148,187,233,1) 100%'
   * 'red, blue'
   * 'rgb(128, 125, 23), rgb(0, 125, 123)'
   */
  gradient?: string;
}

class WebkitInputRangeFillLower {
  /**
   * Formated selectors to use in CSS styles
   * @example
   * #range1, #myrange2
   */
  private _selectors_combined: string;
  /**
   * The dynamic inline style
   * used to store the background values (update progress value)
   */
  private readonly _updatedInlineStyle: HTMLElement;
  /**
   * Class options with List of HTML ID attribute to work with and
   * range elements custom style (background color)
   */
  private _options: WebkitInputRangeFillLowerOptions;
  /**
   * Elements cached values to speed up the Inline Style fill up
   * The system stores in memory the other values, so there is no need
   * to read each value again and perform the calculations
   */
  private _cachedValues: WebkitInputRangeFillLowerCachedValues[];

  /**
   * Webkit input range fill lower hack
   *
   * @param options the range selector ID or an array with multiple IDs
   *                and the range elements custom style (background color)
   *
   * @example
   * document.addEventListener('DOMContentLoaded', () => {
   *    const _test = new WebkitInputRangeFillLower({selectors: ['range1', 'range2'], color: '#ff00ff'});
   *    console.log(_test.getElementsList);
   * });
   *
   * new WebkitInputRangeFillLower({selectors: ['range1', 'range2']}); // with default styles
   * new WebkitInputRangeFillLower({selectors: 'range1', angle: 90, gradient: 'rgba(238,174,202,1) 0%, rgba(198,180,216,1) 74%, rgba(148,187,233,1) 100%'}); // only one element with complex gradient
   */
  constructor(public options: WebkitInputRangeFillLowerOptions) {
    this._updatedInlineStyle = document.createElement('style');
    document.body.append(this._updatedInlineStyle);

    this._cachedValues = [];

    this._validateOptions(options);
    this._removeDuplicates();
    this._removeUnknownElements();
    this._removeInvalidElements();
    this._setSelectorsCombined();
    this._addDefaultStyles();
    this._addEventListeners();
  }

  /**
   * List of ID selectors
   */
  get getElementsList(): string[] {
    return this._options.selectors.map(value => `#${value}`);
  }

  /**
   * Apply default values to invalid settings
   */
  private _validateOptions(options: WebkitInputRangeFillLowerOptions): void {
    const defaultOptions: WebkitInputRangeFillLowerOptions = {
      selectors: null,
      angle: 90,
      color: null,
      gradient: null
    };
    this._options = {...defaultOptions, ...options};

    if (!Array.isArray(options.selectors)) {
      this._options.selectors = [options.selectors];
    }

    if (this._options.angle < 0 || this._options.angle > 360) {
      this._options.angle = 90;
    }

    if (this._options.color === null && this._options.gradient === null) {
      // If none color provided, the default is a gradient
      this._options.gradient = '#0ABFBC, #FC354C';
    }

    if (this._options.color !== null && !this._validateColor(this._options.color)) {
      throw new TypeError('Please enter a valid CSS color');
    }
  }

  /**
   * Add event listeners (input event) to the input range type elements
   */
  private _addEventListeners(): void {
    const eventName = new Event('input');

    for (const id of this._options.selectors) {
      const selector: HTMLElement = document.getElementById(id);

      selector.addEventListener('input', () => {
        const rangeInterval = Number(selector.getAttribute('max')) - Number(selector.getAttribute('min'));
        // @ts-ignore because value
        const rangePercent = (Number(selector.value) + Math.abs(Number(selector.getAttribute('min')))) / rangeInterval * 100;

        // Console.log('#' + id + ': ' + rangePercent + '%');

        this._writeStyle(id, rangePercent);
      }, false);

      selector.dispatchEvent(eventName); // Update the values now
    }
  }

  /**
   * Manage the Cached Data
   * @param id element id
   * @param percent percent value
   */
  private _checkCachedData(id: string, percent: number): void {
    const findResult = this._cachedValues.map(x => x.id).indexOf(id);

    if (findResult === -1) {
      this._cachedValues.push({id, percent});
    } else {
      this._cachedValues[findResult] = {id, percent};
    }
  }

  /**
   * Write the dynamic inline stylesheet with the
   * -webkit-slider-runnable-track CSS pseudo-element
   * Checks for cached values (before writing)
   *
   * @param id element id
   * @param percent percent value
   */
  private _writeStyle(id: string, percent: number): void {
    this._checkCachedData(id, percent);

    let textToWrite = '';

    for (const item of this._cachedValues) {
      textToWrite += `#${item.id}::-webkit-slider-runnable-track{background-size: ${item.percent}% 100% !important }`;
    }

    this._updatedInlineStyle.textContent = textToWrite;
  }

  /**
   * Remove duplicate selectors
   */
  private _removeDuplicates(): void {
    this._options.selectors = this._options.selectors.filter((item, index) => this._options.selectors.indexOf(item) === index);
  }

  /**
   * Remove unknown selectors (invalid IDs)
   */
  private _removeUnknownElements(): void {
    this._options.selectors = this._options.selectors.filter((id: string) => document.getElementById(id) !== null);
  }

  /**
   * Remove elements which aren't an input range type elements
   */
  private _removeInvalidElements(): void {
    // @ts-ignore
    this._options.selectors = this._options.selectors.filter((id: string) => document.getElementById(id).type === 'range');
  }

  /**
   * Stores a list of all IDs combined to use on the dynamic CSS
   */
  private _setSelectorsCombined(): void {
    let selectorText = '';

    this._options.selectors.forEach((element, index) => {
      selectorText += `#${element}`;
      if (index < (this._options.selectors.length - 1)) {
        selectorText += ', ';
      }
    });
    this._selectors_combined = selectorText;
  }

  /**
   * Manage CSS gradients, if only one color is supplied
   * generates a 'solid color' with the same color gradient
   */
  private _generateGradientCSS(): string {
    // Const colors = this._options.color.split(', ');

    let cssColors = '';

    // If (colors.length < 2) {
    //   cssColors = `${this._options.color}, ${this._options.color}`;
    // } else {
    //   cssColors = this._options.color;
    // }

    if (this._options.color === null) {
      cssColors = this._options.gradient; // Use gradient
    } else {
      cssColors = `${this._options.color}, ${this._options.color}`;
    }

    return `${this._options.angle}deg, ${cssColors}`;
  }

  /**
   * Add the default styles to document:
   *  removes default appearence
   *  removes outline
   *  styles the background (webkit-slider-runnable-track) with the custom colors
   *  adds quick fallbacks for Mozilla Firefox and IE
   */
  private _addDefaultStyles(): void {
    let runnableTrackPortion = '';

    this._options.selectors.forEach(id => {
      const backgroundStyle = `background: linear-gradient(${this._generateGradientCSS()}) 0 100% no-repeat content-box !important;`;

      runnableTrackPortion += `#${id}::-webkit-slider-runnable-track {${backgroundStyle}} `;
      runnableTrackPortion += `#${id}::-moz-range-progress {${backgroundStyle}} `;
      runnableTrackPortion += `#${id}::-ms-fill-lower {${backgroundStyle}} `;
    });

    // ${this._selectors_combined} {-webkit-appearance: none !important;}
    const defaultCSS = `<style>
      ${runnableTrackPortion}
      ${this._selectors_combined}:focus {outline: none !important;}
    </style>`;

    document.head.insertAdjacentHTML('beforeend', defaultCSS);
  }

  /**
   * Validate CSS color
   * https://stackoverflow.com/a/48485007/4521967
   * @param strColor color #fff, rgba(255,255,255,1), rgb(255,0,255), etc.
   */
  private _validateColor(stringColor: string): boolean {
    const s = new Option().style;
    s.color = stringColor;
    return s.color !== '';
  }
}
